### ISSUE TITLE
Some questions about the “region_t* extract_requests_*” function

### ISSUE THREAD

**Issue by zan1126** [aflnet/aflnet#16]
Hello, I am a beginner in fuzzing and am very interested in your research. Yesterday, I read about the extract_requests_tls function. One thing is not very clear: 
1、In the function extract_requests_tls (line 242),  you define a variable "bytes_to_skip". I don't quite understand what this variable does? Why skip the payload?
2、Regarding the variables you define in each function of extract_requests_*: byte_count, mem_count, region_count, what do these three variables mean when reading the source code?
I am really interested in your research and hope to get your answers, thank you very much！

**Comment by thuanpv** [comment]
Hi @zan1126,

Thank you for showing your interest in my research. Please find my answers to your questions below

`1、In the function extract_requests_tls (line 242), you define a variable "bytes_to_skip". I don't quite understand what this variable does? Why skip the payload?`

Basically the extract_requests_* function is supposed to just divide the whole sequence of messages/packets (as stored in a seed input) into individual packets so we don't need to distinguish between the message header and the payload; as long as we can extract individual messages, it would be fine. Having said that, we do have a plan to improve AFLNet to make it message structure-aware and apply some techniques that we proposed in our previous smart grey-box fuzzing work (see AFLSmart at https://github.com/aflsmart/aflsmart).

`2、Regarding the variables you define in each function of extract_requests_*: byte_count, mem_count, region_count, what do these three variables mean when reading the source code?`

byte_count is used to keep track of how many bytes the function has read from the initial seed input. The initial seed input is supposed to keep the whole sequence of messages sent from a client to the server under test. Please read README.md and see how a seed file to fuzz the Live555 RTSP server is prepared.

mem_count keeps track of how many bytes the function has used to store intermediate data for the current region/message. This variable is reset once the function reaches the end of a message. 

region_count keeps track of the number of messages/regions the function so far has extracted from the seed input. In AFLNet I use region & message terminologies together. At the beginning 1 region = 1 valid message; however, during fuzzing 1 region could contain more than 1 (invalid) message respect to the message format.

I hope my answers help.

Best regards,

Thuan

**Comment by zan1126** [comment]
Thank you very much for your answers @thuanpv 
When reading the source code, I had previously misunderstood the extract_requests_* function, and mistakenly thought that it was necessary to distinguish the message header from the payload and only extract the payload part. And I have a good understanding of the three variables in each function.

But , as I continued to read the source code, several troublesome questions appeared, and I hope to get your answers:
1、What do the three parameters buf, buf_size, region_count_ref/state_count_ref in function extract_requests_* or extract_response_codes* mean? At present, what I understand is that buf represents the content of the message sequence to be extracted, buf_size represents its size, and region_count_ref does not understand what it means. I don’t know if I understand it or not?
2、When I read the source code of function xxx, I did not understand what this function is responsible for? Is it responsible for extracting the server's response code? In your paper, it is mentioned that “the FTP server responds with a response message containing a status code”, I would like to ask if the status code is not included in the response message of the server of the protocol, how should the response code be extracted? In other words, the response code of any protocol server includes the status code?

I am studying the source code of your project. There may be many questions about this and I need to ask you for advice. Thank you very much for taking the time out of your busy schedule to answer the "non-nutritive" questions I asked. Your answers have promoted my understanding of the entire project. Thank you very much!

**Comment by thuanpv** [comment]
Hi @zan1126 

Regarding the functions extract_requests_*, please find below the explanation for the input arguments
- buf: a pointer to the memory region keeping a sequence of message/packets -- the extract_requests_** function splits this sequence to a list of regions/messages
- buf_size: the size of buf
- region_count_ref: keeps the number of regions/messages/packets found in buf

Regarding the extract_response_codes_**:
- buf: a pointer to the memory region keeping the server responses -- sent by the server under test while consuming the messages sent from AFLNet; AFLNet plays the role of a client.
- buf_size: size of buf
- state_count_ref: keeps the number of states found in buf
This function returns a list of state IDs.

`I would like to ask if the status code is not included in the response message of the server of the protocol, how should the response code be extracted? In other words, the response code of any protocol server includes the status code?`

If the status code is not explicitly included in the server responses, you could consider to calculate a kind-of hash value based on some part(s) of the server response and these hash values can be used as response codes. For example, for TLS protocol, I generate the hash value based on two different values stored in the server responses -- message_type and content_type (https://github.com/aflnet/aflnet/blob/master/aflnet.c#L734). 

I hope it helps.

Thuan

**Comment by zan1126** [comment]
Thank you very much, if I encounter other problems or come up with different methods, I will discuss with you here, thank you. @thuanpv

**Comment by zan1126** [comment]
Hi, @thuanpv 
Through the previous communication with you, I have a preliminary understanding of AFLNET and try to expand it. I added another network protocol to AFLNET. Based on my understanding of the network protocol and the content in aflnet.c, I wrote the two-part code for extracting the request and response for the protocol, and added it to afl-fuzz.c. The information of the agreement is added everywhere. But when I started the experiment and started the test, a problem called "segment fault" appeared, which directly caused the AFLNET test result panel to not be displayed. I have also tried to modify, but this problem appeared. Would like to ask you what is causing this problem? Or is it because the two parts of the code are incorrectly written? Please reply me at your convenience, thank you very much!

**Comment by thuanpv** [comment]
Hi @zan1126. Can you send me the link to your code and share with me the instructions to test a server implementing the protocol under test?

**Comment by zan1126** [comment]
Sorry, @thuanpv ,I re-learned your source code and modified the relevant code I wrote, so some time was lost. The following are two parts of the code I wrote, I hope to get your guidance to find out the mistakes, thank you very much.
The codes are in https://github.com/zan1126/aflnet-new-protocol

**Comment by zan1126** [comment]
Sorry, @thuanpv .There were some problems with the previous link. After I modified the code, I put it in this link. Thank you very much for your guidance. https://github.com/zan1126/aflnet-add-protocol/tree/master

**Comment by zan1126** [comment]
Hi, @thuanpv 
I added my sample PCAP file capturing the traffic between ngtcp2 client and server, it is called ngtcp2.pcap. In my opinion, I think it includes stateful information in the response message, and I also wrote extract_response.c function to extract stateful information. But when I run ngtcp2 in AFLNET, there will be a "segmentation" fault. So, I removed the -E -q -s options based on your previous answers to other questions, and run it again. The operation is successful, and the results of its stage operation are shown in the figure. I don't know the reason, because this protocol is stateful (in my opinion). But the testing was successful when not using extract_reaponse.c function. 

So, I ask you to take a look at my sample file and check whether the response information of the file contains status information? If it contains, why is there a segfault when applying the extract_response.c function? I really look forward to your help, thank you again!

The above mentioned PCAP file and test results are in https://github.com/zan1126/aflnet-add-protocol

Thank you very much!!!!
